core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4QueryHttpSigner.java

```diff
@@ -31,11 +57,133 @@ public final class DefaultAwsV4QueryHttpSigner implements AwsV4QueryHttpSigner {
+                normalizePath);
+
+        String signingCredentials = sanitizedCredentials.accessKeyId() + "/" + scope;
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_ALGORITHM, SignerConstant.AWS4_SIGNING_ALGORITHM);
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_DATE, formattedRequestSigningDateTime);
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_SIGNED_HEADERS, canonicalRequest.signedHeaderString());
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_EXPIRES, Long.toString(expirationInSeconds));
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_CREDENTIAL, signingCredentials);
+
+        String canonicalRequestString = canonicalRequest.string();
+        String stringToSign = buildStringToSign(
+            canonicalRequestString,
+            AWS4_SIGNING_ALGORITHM,
+            formattedRequestSigningDateTime,
+            scope
+        );
+
+        byte[] signingKey = deriveSigningKey(
+            sanitizedCredentials,
+            requestSigningInstant,
+            regionName,
+            serviceSigningName
+        );
+
+        byte[] signature = computeSignature(stringToSign, signingKey);
+
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_SIGNATURE, BinaryUtils.toHex(signature));
+
+        return requestBuilder;
+    }
+
+    private void addSessionCredentials(SdkHttpRequest.Builder requestBuilder,
+                                       AwsSessionCredentialsIdentity credentials) {
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_SECURITY_TOKEN, credentials.sessionToken());
     }
 
+
     @Override
     public AsyncSignedRequest signAsync(AsyncSignRequest<? extends AwsCredentialsIdentity> request) {
-        throw new UnsupportedOperationException();
+        if (CredentialUtils.isAnonymous(request.identity())) {
+            return AsyncSignedRequest.builder()
+                .request(request.request())
+                .payload(request.payload().orElse(null))
+                .build();
+        }
+
+        DigestComputingSubscriber bodyDigester = DigestComputingSubscriber.forSha256();
+
+        request.payload().ifPresent((payload) ->
+            payload.subscribe(bodyDigester)
+        );
+
+        CompletableFuture<byte[]> digestBytes = bodyDigester.digestBytes();
+
+        CompletableFuture<SdkHttpRequest> signedReqFuture = digestBytes.thenApply(bodyHash -> {
+            String digestHex = BinaryUtils.toHex(bodyHash);
+
+            SdkHttpRequest.Builder builder = doSign(request, digestHex);
+
+            return builder.build();
+        });
+
+        return AsyncSignedRequest.builder()
+            .request(CompletableFutureUtils.joinLikeSync(signedReqFuture))
+            .payload(request.payload().orElse(null))
+            .build();
+    }
+
+    /**
+     * Generates an expiration time, in seconds, for the presigned url. If user has specified
```

Comment:
"Generates" - this method does not need to generate/return anything. It's just a validation method that can return void. The caller already has the Duration and can convert there.